File: 39.组合总和.go
/*
 * @lc app=leetcode.cn id=39 lang=golang
 *
 * [39] 组合总和
 */

// @lc code=start
func combinationSum(candidates []int, target int) (ret [][]int) {
	var backtrack func([]int, int, int)
	backtrack = func (item []int, remain int, idx int) {
		if idx >= len(candidates) {
			return
		}
		if remain == 0 {
			newItem := append([]int{}, item...)
			ret = append(ret, newItem)
			return
		}

		backtrack(item, remain, idx+1)
		
		if remain - candidates[idx] >= 0 {
			newItem := append([]int{}, item...)
			backtrack(append(newItem, candidates[idx]), remain-candidates[idx], idx)
		}
	}
	backtrack([]int{}, target, 0)
	return
}
// @lc code=end

File: 39.组合总和.ts
/*
 * @lc app=leetcode.cn id=39 lang=typescript
 *
 * [39] 组合总和
 */

// @lc code=start
function combinationSum(candidates: number[], target: number): number[][] {
    let ret: number[][] = [];
    function backtrack(item: number[], remain: number, idx: number) {
        if (idx === candidates.length) return;
        if (remain === 0) {
            ret.push(item);
            return;
        }
        backtrack(item, remain, idx+1);

        if (remain - candidates[idx] >= 0) {
            backtrack([...item, candidates[idx]], remain - candidates[idx], idx);
        }
    }
    backtrack([], target, 0);
    return ret;
};
// @lc code=end

